src/lib/pubkey/dilithium/dilithium_common/dilithium_constants.h

```diff
@@ -0,0 +1,171 @@
+/*
+ * Crystals Dilithium Constants
+ *
+ * (C) 2022-2023 Jack Lloyd
+ * (C) 2022      Manuel Glaser - Rohde & Schwarz Cybersecurity
+ * (C) 2022-2023 Michael Boric, René Meusel - Rohde & Schwarz Cybersecurity
+ * (C) 2024 René Meusel, Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#ifndef BOTAN_DILITHIUM_CONSTANTS_H_
+#define BOTAN_DILITHIUM_CONSTANTS_H_
+
+#include <botan/dilithium.h>
+
+namespace Botan {
+
+class Dilithium_Symmetric_Primitives;
+
+/**
+ * Algorithm constants and parameter-set dependent values
+ */
+class DilithiumConstants final {
+   public:
+      /// base data type for most calculations
+      using T = int32_t;
+
+      /// number of coefficients in a polynomial
+      static constexpr T N = 256;
+
+      /// modulus
+      static constexpr T Q = 8380417;
+
+      /// number of dropped bits from t (see FIPS 204 Section 5)
+      static constexpr T D = 13;
+
+      /// as specified in FIPS 204 (see Algorithm 36 (NTT^-1), f = 256^-1 mod Q)
+      static constexpr T F = 8347681;
+
+      /// the 512-th root of unity modulo Q (see FIPS 204 Section 8.5)
+      static constexpr T ROOT_OF_UNITY = 1753;
+
+      /// degree of the NTT polynomials
+      static constexpr size_t NTT_Degree = 256;
+
+   public:
+      /// \name Byte length's of various hash outputs and seeds
+      /// @{
+
+      static constexpr size_t SEED_RANDOMNESS_BYTES = 32;
+      static constexpr size_t SEED_RHO_BYTES = 32;
+      static constexpr size_t SEED_RHOPRIME_BYTES = 64;
+      static constexpr size_t SEED_SIGNING_KEY_BYTES = 32;
+      static constexpr size_t MESSAGE_HASH_BYTES = 64;
+      static constexpr size_t PUBLIC_KEY_HASH_BYTES = 32;
+      static constexpr size_t COMMITMENT_HASH_FULL_BYTES = 32;
+      static constexpr size_t COMMITMENT_HASH_C1_BYTES = 32;
+
+      /// @}
+
+   public:
+      enum DilithiumTau : uint32_t { _39 = 39, _49 = 49, _60 = 60 };
+
+      enum DilithiumLambda : uint32_t { _128 = 128, _192 = 192, _256 = 256 };
+
+      enum DilithiumGamma1 : uint32_t { ToThe17th = (1 << 17), ToThe19th = (1 << 19) };
+
+      enum DilithiumGamma2 : uint32_t { Qminus1DevidedBy88 = (Q - 1) / 88, Qminus1DevidedBy32 = (Q - 1) / 32 };
+
+      enum DilithiumEta : uint32_t { _2 = 2, _4 = 4 };
+
+      enum DilithiumBeta : uint32_t { _78 = 78, _196 = 196, _120 = 120 };
+
+      enum DilithiumOmega : uint32_t { _80 = 80, _55 = 55, _75 = 75 };
```

Comment:
Dilithium has many integer constants that change depending on the instantiation of the algorithm. I found it worthwhile to restrict them with explicit enums. I'm okay with this concept, also given the restricted (algorithm-internal) scope. But I feel there must be a better way for this. :)

Reply:
I also like this idea. While the implied switch-cases are annoying, they guarantee we do not forget a value.